utils/progress.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2016 Mender Software AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+package utils
+
+import (
+	"fmt"
+	"io"
+)
+
+type ProgressWriter struct {
+	Out io.Writer // progress output
+	N   int64     // size of the input
+	c   int64     // current count
+}
+
+func (p *ProgressWriter) Write(data []byte) (int, error) {
+	n := len(data)
+
+	p.reportGeneric(n)
+	p.c += int64(n)
+	return n, nil
+}
+
+func (p *ProgressWriter) reportGeneric(n int) {
+	const perLine = 1 * 1024 * 1024 // 1MB
+	const dotsPerLine = 32
+	const perDot = perLine / dotsPerLine
+
+	now := p.c
+	then := p.c + int64(n)
+	nowDots := now / perDot
+	thenDots := then / perDot
+
+	for ; nowDots < thenDots; nowDots++ {
+		p.Out.Write([]byte("."))
+		if nowDots != 0 && (nowDots+1)%dotsPerLine == 0 {
+			// print percentage and size after each complete line
+			var s string
+			nowSize := (nowDots + 1) * perDot
+			nowSizekB := nowSize / 1024
+			if p.N == 0 {
```

Comment:
I would perhaps do:

```
if p.N == 0 || then > p.N
```

to catch errors where the expected size is miscalculated. Unlikely, but possible. We might consider a warning as well.